<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [61]:
import pandas as pd
import psycopg2

In [62]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [63]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query = f'''
select
COUNT(id)
from public.VACANCIES
'''

In [5]:
# результат запроса
df = pd.read_sql_query(query, connection)
display(df)

count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query = f'''
select
COUNT(id)
from public.EMPLOYERS
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query, connection)
display(df)

count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query = f'''
select
COUNT(id)
from public.AREAS
'''

In [9]:
# результат запроса
df = pd.read_sql_query(query, connection)
display(df)

count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query = f'''
select
COUNT(id)
from public.INDUSTRIES
'''

In [11]:
# результат запроса
df = pd.read_sql_query(query, connection)
display(df)

count
0    294

***

### выводы по предварительному анализу данных

Кол-во вкансий составляет 49197, их вдвое больше чем работадалтелей, что весьма логично так как на одну компанию может приходится несколько вакансий, кол-во регионов и сфер деятельности не дало пока общего понимания исходя из цыфр, данные связаны между собой по ID что позволит выявлять различные закономерности и связи для более глубокого анализа данных.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''SELECT 
    a.name as areas_name,
    COUNT(v.id) as "Кол-во вакансий"
FROM   
    public.areas a
JOIN public.VACANCIES v ON v.area_id = a.id
GROUP BY areas_name
ORDER BY 2 DESC
LIMIT 5
'''

In [13]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
display(df)

areas_name  Кол-во вакансий
0           Москва             5333
1  Санкт-Петербург             2851
2            Минск             2112
3      Новосибирск             2006
4           Алматы             1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''SELECT 
    COUNT(VACANCIES.id) as "Кол-во вакансий"
FROM   
    public.VACANCIES
    WHERE salary_from IS NOT NULL or salary_to IS NOT NULL
'''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
display(df)

Кол-во вакансий
0            24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''SELECT 
    round (AVG(salary_from)) AS "средняя по нижней границе зп",
    round (AVG(salary_to)) AS "средняя по верхней границе зп"
FROM   
    public.VACANCIES
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
display(df)

средняя по нижней границе зп  средняя по верхней границе зп
0                       71065.0                       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''SELECT 
    VACANCIES.schedule as "Тип рабочего графика",
    VACANCIES.employment as "Тип трудоустройства",
    COUNT(VACANCIES.id) as "Количество вакансий"
FROM   
    public.VACANCIES
GROUP BY VACANCIES.schedule,VACANCIES.employment
ORDER BY 3 DESC
LIMIT 5
'''

In [19]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
display(df)

Тип рабочего графика  Тип трудоустройства  Количество вакансий
0          Полный день     Полная занятость                35367
1     Удаленная работа     Полная занятость                 7802
2        Гибкий график     Полная занятость                 1593
3     Удаленная работа  Частичная занятость                 1312
4       Сменный график     Полная занятость                  940

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''SELECT 
    VACANCIES.experience as "Опыт работы",
    COUNT(VACANCIES.id) as "Кол-во вакансий"
FROM   
    public.VACANCIES
GROUP BY VACANCIES.experience
ORDER BY 2 ASC
LIMIT 4
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
display(df)

Опыт работы  Кол-во вакансий
0         Более 6 лет             1337
1           Нет опыта             7197
2       От 3 до 6 лет            14511
3  От 1 года до 3 лет            26152

***

### выводы по детальному анализу вакансий
По предварительному анализу видно что спрос на вкансии сконцентрирован вокруг городов миллионников, наибольшое кол-во вакансий хотят чтобы соискатели  работатали полный день и на полную знатятость, следом идут удаленка и гибкий граффик, что говорит потенциале роста в этих двух сферах в будущем при росте вакансий не завязанных на постоянную работу в офисе, средние ожидание по зп от 71 до 110,  наибольшая концентрация по опыту в  cуммарно двух диапзонах 1"От 1 года до 3 лет", 2'От 3 до 6 лет'.
Логично что меньше вакансий  с опытом "Более 6 лет", так как там и требования выше чем во всех остальных.


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''(SELECT 
    EMPLOYERS.name as "Работодатель",
    COUNT(VACANCIES.id) as "Кол-во вакансий"
FROM   
    public.EMPLOYERS
JOIN public.VACANCIES ON VACANCIES.employer_id = EMPLOYERS.id
GROUP BY EMPLOYERS.name
ORDER BY 2 DESC
LIMIT 1)
UNION ALL
(SELECT 
    EMPLOYERS.name as "Название работодателя",
    COUNT(VACANCIES.id) as "Кол-во вакансий"
FROM   
    public.EMPLOYERS
JOIN public.VACANCIES ON VACANCIES.employer_id = EMPLOYERS.id
GROUP BY EMPLOYERS.name
ORDER BY 2 DESC
OFFSET 4 LIMIT 11)
'''

In [23]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
display(df)

Работодатель  Кол-во вакансий
0                  Яндекс             1933
1           Газпром нефть              331
2             ИК СИБИНТЕК              327
3                     МТС              292
4                 DataArt              247
5   Совкомбанк Технологии              204
6              Первый Бит              176
7                    Ozon              175
8                     IBS              175
9                Andersen              155
10                 Тензор              155
11             Спецремонт              151

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = f'''SELECT 
    a.name as "Регион",
    COUNT(distinct e.id) as "Кол-во работодателей",
    COUNT(distinct v.id) as "Кол-во вакансий"
FROM   
    public.AREAS a
JOIN public.EMPLOYERS e ON e.area = a.id
LEFT JOIN public.VACANCIES v ON v.area_id = a.id
WHERE v.id is NULL
GROUP BY a.name
ORDER BY 2 DESC
LIMIT 1
'''

In [25]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
display(df)

Регион  Кол-во работодателей  Кол-во вакансий
0  Россия                   410                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3 = f'''SELECT 
    e.name as "Работодатель",
    COUNT(distinct v.area_id) as "Кол-во регионов публикующие вакансии",
    COUNT(distinct v.id) as "Кол-во вакансий"
FROM   
    public.EMPLOYERS e
JOIN public.VACANCIES v ON v.employer_id = e.id
GROUP BY e.name
ORDER BY 3 DESC
LIMIT 60
'''

In [27]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
display(df)

Работодатель  Кол-во регионов публикующие вакан  \
0                                  Яндекс                                181   
1                              Ростелеком                                152   
2                                Тинькофф                                 43   
3                                    СБЕР                                 24   
4                           Газпром нефть                                 24   
5                             ИК СИБИНТЕК                                 46   
6                                     МТС                                 55   
7                                 DataArt                                  2   
8                   Совкомбанк Технологии                                 26   
9                              Первый Бит                                 43   
10                                   Ozon                                 36   
11                                    IBS                                 36   
12                                 Тензор                                 25   
13                               Andersen                                 15   
14                             Спецремонт                                116   
15                                     VK                                 17   
16                 Поляков Денис Иванович                                 88   
17               Иннотех, Группа компаний                                 19   
18                        Bell Integrator                                 26   
19                                  ANCOR                                 44   
20                                  ЛАНИТ                                 29   
21                 МАГНИТ, Розничная сеть                                 25   
22                Лига Цифровой Экономики                                 15   
23                               Softline                                 25   
24                                МегаФон                                 23   
25                             Совкомбанк                                 63   
26               Центр Высоких Технологий                                 15   
27                         Банк ВТБ (ПАО)                                 24   
28                               ИнфоТеКС                                  6   
29             ЭФКО, Управляющая компания                                 49   
30                                Playrix                                 25   
31  Центральный банк Российской Федерации                                 19   
32                     Nitka Technologies                                  7   
33                               1C-Рарус                                 28   
34                                 Контур                                 16   
35                              Норникель                                 20   
36                             СимбирСофт                                 19   
37                           Почта России                                 48   
38                               Гринатом                                 20   
39                  Positive Technologies                                  6   
40                      DNS Головной офис                                  6   
41                            Спортмастер                                 25   
42                              ООО ЕФИН                                  71   
43                                  АЛНАС                                 41   
44                             Северсталь                                 21   
45                                 Диплей                                 11   
46                                  Tele2                                 11   
47                               SberTech                                 13   
48                             ЭР-Телеком                                 33   
49                           

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4 = f'''SELECT
    COUNT(distinct e.id) as "Кол-во работодателей"
FROM   
    public.EMPLOYERS e
LEFT JOIN public.EMPLOYERS_INDUSTRIES i ON e.id = i.employer_id
WHERE i.industry_id IS NULL
'''

In [29]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
display(df)

Кол-во работодателей
0                  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_5_5 = f'''SELECT
    e.name as "Компания",
    COUNT(distinct i.industry_id) as "Кол-во сфер деятельности"
FROM   
    public.EMPLOYERS e
LEFT JOIN public.EMPLOYERS_INDUSTRIES i ON e.id = i.employer_id
GROUP BY e.name
HAVING COUNT(distinct i.industry_id) = 4
ORDER BY 1 ASC
OFFSET 2 LIMIT 1
'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
display(df)

Компания  Кол-во сфер деятельности
0     2ГИС                         4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''SELECT
    COUNT(distinct e.id) as "Кол-во работодателей"
FROM   
    public.EMPLOYERS e   
LEFT JOIN public.EMPLOYERS_INDUSTRIES i ON e.id = i.employer_id
JOIN public.INDUSTRIES y ON y.id = i.industry_id 
WHERE y.name = 'Разработка программного обеспечения'
'''

In [33]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
display(df)

Кол-во работодателей
0                  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников
from urllib.parse import quote_plus
url = quote_plus('https://ru.wikipedia.org/wiki/Города-миллионеры_России', safe=':/')
tables = pd.read_html(url)
mln_cities = tables[1]['Город'].values.tolist()

In [35]:
# текст запроса
query_5_7 = f'''
    select
        a.name as region,
        count(*) as vacancies_count
    from
        public.vacancies as v
        join public.employers as e on v.employer_id = e.id
        join public.areas as a on v.area_id = a.id
    group by
        v.area_id, a.name, e.id, e.name
    having
        a.name in {tuple(mln_cities)} and e.name = 'Яндекс'
    

    union all
    

    select
        'Total',
        sum(x.cnt)
    from
        (
        select
            count(*) as cnt
        from
            public.vacancies as v
            join public.employers as e on v.employer_id = e.id
            join public.areas as a on v.area_id = a.id
        group by
            v.area_id, a.name, e.id, e.name
        having
            a.name in {tuple(mln_cities)} and e.name = 'Яндекс'
        ) as x
        

    order by 2
    '''

In [36]:
# результат запроса
df = pd.read_sql_query (query_5_7, connection)
df

region  vacancies_count
0              Омск             21.0
1         Челябинск             22.0
2        Красноярск             23.0
3         Волгоград             24.0
4    Ростов-на-Дону             25.0
5            Казань             25.0
6             Пермь             25.0
7               Уфа             26.0
8            Самара             26.0
9         Краснодар             30.0
10          Воронеж             32.0
11      Новосибирск             35.0
12  Нижний Новгород             36.0
13     Екатеринбург             39.0
14  Санкт-Петербург             42.0
15           Москва             54.0
16            Total            485.0

***

### выводы по анализу работодателей
Большинство курпных работадатлей находятся в сферах ИТ,Телекомуникаций,Финансы, традиционно большинство вакансий сосредоточенно вокруг городов милионников, там где насленеие больше 1 млн человек наблюдается наибольшее кол-во вакансий

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [64]:
# текст запроса
query_6_1 = '''select count(*)
                from vacancies
                where name ilike '%data%' or name ilike '%данн%'
'''

In [65]:
# результат запроса
df = pd.read_sql_query (query_6_1, connection)
df

count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [82]:
# текст запроса
query_6_2 = '''select count(*)
    from
        public.vacancies v
    where
        (v.name ILIKE '%data scientist%' OR
        v.name ILIKE '%data science%'OR
        v.name ILIKE '%исследователь данных%' OR
        (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
        v.name ILIKE '%machine learning%' OR v.name ILIKE '%машинн%обучен%')
        and
        (v.name ILIKE '%junior%' or
        v.experience = 'Нет опыта' or
        v.employment ILIKE '%стажировка%')
'''
  

In [83]:
# результат запроса
df = pd.read_sql_query (query_6_2, connection)
df

count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [94]:
# текст запроса
query_6_3 = f''' select
count(*)
from vacancies v
where(v.name ilike '%data scientist%' 
or v.name ilike '%data science%' 
or v.name ilike '%data science%'
 or v.name ilike '%исследователь данных%'
 or( v.name like '%ML%' and v.name not like '%HTML%')
 or v.name ilike '%machine learning%'
 or v.name ilike '%машинн%обучен%')
and (v.key_skills ilike '%SQL%' or  v.key_skills ilike '%postgres%')
'''

In [95]:
# результат запроса
df = pd.read_sql_query (query_6_3, connection)
df

count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [126]:
# текст запроса
query_6_4 = f'''
select count(*)
    from
        public.vacancies
    where
        (
        lower(name) like '%data scientist%'  or
        lower(name) like '%data science%' or
        lower(name) like '%исследователь данных%'  or
        (name like '%ML%' and name not ilike '%HTML%')  or
        lower(name) like '%machine learning%'  or
        lower(name) like '%машинн%обучен%'
        )
        and
        key_skills ilike '%Python%'
'''

In [127]:
# результат запроса
df = pd.read_sql_query (query_6_4, connection)
df

count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [136]:
# текст запроса
query_6_5 = f'''
SELECT 
    round(avg(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1), 2) as result
FROM public.VACANCIES v
where 
(
  name ilike '%data scientist%'or
  name ilike '%data science%'or
  name ilike '%исследователь данных%'or
  (name like '%ML%' and name not ilike '%HTML%') or
  name ilike '%machine learning%' or name ilike '%машинн%обучен%'
)
'''

In [137]:
# результат запроса
df = pd.read_sql_query (query_6_5, connection)
df

result
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [151]:
# текст запроса
query_6_6 = f'''
select
distinct(v.experience) ex, avg(coalesce(((v.salary_from+v.salary_to)/2),v.salary_from,v.salary_to)) avg_salary_total
from vacancies v
where (name ilike '%data scientist%'
       or name ilike '%data science%'
       or name ilike '%исследователь данных%'
       or (name like '%ML%' and name not ilike '%HTML%')
       or name ilike '%machine learning%' or name ilike '%машинн%обучен%'
       )
       AND experience = 'От 3 до 6 лет'
AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY experience
'''

In [152]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

ex  avg_salary_total
0  От 3 до 6 лет     243114.666667

***

### выводы по предметному анализу
В предметном анализе мы видим взаимосвязь между количеством навыков а также опыта и зависимостью их от потенциальной зп

# Общий вывод по проекту

В целом благодоря взаимосвязи таблиц по ID довольно чётко можно выявить закономерности в разных разрезах,с учетом вакансий, отраслей работодателей, городов, а также опыта и предпологаемых вилок зп.
Так же видно что при прочих равных специалисты по DS и ML востребованны на рынке труда и основываясь на личном опыте зп у них на 20-30 % выше в сравнении с обычным разработчиком, при сходном опыте и навыках.

